### PR TITLE
Support Spark over Gloo

### PR DESCRIPTION
**Pull request by EnricoMi**
This adds Gloo-support for Horovod on Spark. This PR might clash with PR #1722.

### CODE REVIEW

**Comment by EnricoMi** [comment]
@tgaddair I am stuck with a unit test failure:

The `test_spark.py::SparkTests::test_happy_run_with_mpi` unit test fails with MPICH in `test-cpu-mpich-py3_6-tf1_14_0-keras2_3_1-torch1_3_0-mxnet1_5_0-pyspark2_4_0_`: [log](https://buildkite.com/organizations/horovod/pipelines/horovod/builds/2259/jobs/edf13a2a-00d7-47e3-990e-bfd96552686a/artifacts/0c02b123-3f95-44d2-ba8a-fb224c60861a).

So `mpirun` does not understand `--allow-run-as-root`. In `Dockerfile.test.cpu` the `mpirun` command for MPICH is `mpirun -np 2` only, whereas for OpenMPI it contains `-allow-run-as-root`: `mpirun -allow-run-as-root -np 2 -H localhost:2 -bind-to none -map-by slot -mca mpi_abort_print_stack 1`. The options `-allow-run-as-root` and `--allow-run-as-root` are synonyms.

Btw., `mpirun` rightly says:
> Allow execution as root (STRONGLY DISCOURAGED)

Shouldn't this be removed? Users can add it to the mpi extra args.

There are other MPICH docker images that succeed. The `mpirun` command errors with:
```
[mpiexec@a46ecacae1f8] match_arg (utils/args/args.c:159): unrecognized argument allow-run-as-root         
[mpiexec@a46ecacae1f8] HYDU_parse_array (utils/args/args.c:174): argument matching returned error         
[mpiexec@a46ecacae1f8] parse_args (ui/mpich/utils.c:1597): error parsing input array                      
[mpiexec@a46ecacae1f8] HYD_uii_mpx_get_parameters (ui/mpich/utils.c:1649): unable to parse user arguments 
[mpiexec@a46ecacae1f8] main (ui/mpich/mpiexec.c:149): error parsing parameters                            
[mpiexec@a46ecacae1f8] match_arg (utils/args/args.c:159): unrecognized argument allow-run-as-root         
[mpiexec@a46ecacae1f8] HYDU_parse_array (utils/args/args.c:174): argument matching returned error         
[mpiexec@a46ecacae1f8] parse_args (ui/mpich/utils.c:1597): error parsing input array                      
[mpiexec@a46ecacae1f8] HYD_uii_mpx_get_parameters (ui/mpich/utils.c:1649): unable to parse user arguments 
[mpiexec@a46ecacae1f8] main (ui/mpich/mpiexec.c:149): error parsing parameters                            
```

I cannot see why this should not have failed before (i.e. in master) given my changes do not touch the executed `mpirun`. Further other MPICH docker images work fine. I am running out of ideas investigating this.

Subsequent tests seem to have been cancelled by buildkite due to the failure. Is there a way to make buildkite not failfast but run all tests so I can see the full picture of issues with one run?

**Comment by tgaddair** [comment]
Hey @EnricoMi, to answer your questions: previously, Spark tests were passing because they were only run on OpenMPI.  I saw in your diff for `test_spark.py` that this line was removed:

```
if 'Open MPI' not in str(subprocess.check_output('mpirun --version', shell=True)):	
            return
```

So as a result, Spark tests are now running with MPICH.  One option for now is to skip Spark tests when MPICH is used.  You can add something like this to `run_mpi_pytest()` in `gen-pipeline.sh`:

```
local exclude_spark_if_needed=""
  if [[ ${test} == *"mpich"* ]]; then
    exclude_spark_if_needed="| sed 's/[a-z_]*spark[a-z_.]*//g'"
  fi

run_test ... ${exclude_keras_if_needed} ${exclude_interactiverun} ${exclude_spark_if_needed}
```

We can address the issue of making Horovod on Spark compatible with MPICH in a separate PR.  I think `--allow-run-as-root` is useful when running in containerized environments, where you frequently run as a root user (but because you're in a container, your ability to break things is very limited).

It looks like the tests that didn't run are GPU tests.  Since your changes don't affect those code paths, I wouldn't expect them to fail, but one way to make sure they run is to remove the `echo "- wait"` lines in `gen-pipeline.sh`.  But in general I don't recommend this, as GPU tests cost more to run, so we try to make sure all CPU tests pass first before running them.

**Comment by EnricoMi** [comment]
Excellent spot, I was wondering what that line was used for ;-)

I removed it because most of the tests don't use MPI at all even though they test MPI code paths. They run on platforms where no MPI or Gloo are available. But I know what to do now. It got late yesterday so I missed that obviously, thanks!

**Comment by EnricoMi** [comment]
@tgaddair tests look much better now. I have activated Spark unit tests and integration test for non-mpi environments (i.e. gloo) for both, cpu and gpu. Here is an example: https://buildkite.com/horovod/horovod/builds/2277#b0ff80ef-ed82-4294-b2ef-ae2af6866540/210-529

So it looks like Spark works over Gloo.

There are a few gpu tests that fail first run and succeed later. I cannot see a connection with Spark over Gloo:
https://buildkite.com/horovod/horovod/builds/2277#37070f44-46df-4dfa-8f20-ac31a210b018/306-491
```
2020-03-22T18:21:42.900603859Z [1,0]<stderr>:tensorflow.python.framework.errors_impl.UnknownError: 2 root error(s) found.
2020-03-22T18:21:42.900606115Z [1,0]<stderr>:  (0) Unknown: Failed to get convolution algorithm. This is probably because cuDNN failed to initialize, so try looking to see if a warning log message was printed above.
2020-03-22T18:21:42.900608586Z [1,0]<stderr>:	 [[{{node conv2d/Conv2D}}]]
2020-03-22T18:21:42.900610690Z [1,0]<stderr>:	 [[loss/mul/_119]]
2020-03-22T18:21:42.900612832Z [1,0]<stderr>:  (1) Unknown: Failed to get convolution algorithm. This is probably because cuDNN failed to initialize, so try looking to see if a warning log message was printed above.
2020-03-22T18:21:42.900615288Z [1,0]<stderr>:	 [[{{node conv2d/Conv2D}}]]
```

https://buildkite.com/horovod/horovod/builds/2277#8b4b02c3-8a65-40f4-ad4e-818978919965/210-377
```
2020-03-22 18:23:16.003652: E tensorflow/stream_executor/cuda/cuda_dnn.cc:329] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
2020-03-22 18:23:16.010340: E tensorflow/stream_executor/cuda/cuda_dnn.cc:329] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
```

https://buildkite.com/horovod/horovod/builds/2277#e28449bc-7892-4579-9c37-ef146244cacd/211-242
```
test_spark_torch.py::SparkTorchTests::test_fit_model FAILED
```

Is this flakiness usual?

Then there are two gpu tests (PyTests Spark Estimators with gloo and openmpi) that fail all three times:
- test-gpu-openmpi-py3_6-tf1_15_0-keras2_3_1-torch1_3_0-mxnet1_4_1-pyspark2_4_0
- test-gpu-gloo-py3_6-tf1_15_0-keras2_3_1-torch1_3_0-mxnet1_4_1-pyspark2_4_0

https://buildkite.com/horovod/horovod/builds/2277#37ce0a26-e05b-49df-81ca-83cff087c5a3
```
test_spark_keras.py::SparkKerasTests::test_fit_model FAILED
test_spark_keras.py::SparkKerasTests::test_fit_model_multiclass FAILED
test_spark_keras.py::SparkKerasTests::test_model_serialization FAILED
```

https://buildkite.com/horovod/horovod/builds/2277#2e5f8ab3-bee8-4dc2-a0ef-a5daf2a3c0df
```
test_spark_keras.py::SparkKerasTests::test_fit_model PASSED
test_spark_keras.py::SparkKerasTests::test_fit_model_multiclass FAILED
test_spark_keras.py::SparkKerasTests::test_model_serialization PASSED
```

They fail on `run_results = [None, None]`:
```
2020-03-22T18:26:19.854924576Z self = KerasEstimator_c723c3016bdf, run_results = [None, None]
2020-03-22T18:26:19.854927192Z run_id = 'keras_1584901382'
2020-03-22T18:26:19.854930254Z metadata = {'features': {'intermediate_format': 'nochange', 'is_sparse_vector_only': False, 'max_size': None, 'shape': None, ...}, 'y': {'intermediate_format': 'nochange', 'is_sparse_vector_only': False, 'max_size': None, 'shape': None, ...}}
2020-03-22T18:26:19.854934027Z 
2020-03-22T18:26:19.854936988Z     def _create_model(self, run_results, run_id, metadata):
2020-03-22T18:26:19.854940088Z         keras_utils = self._get_keras_utils()
2020-03-22T18:26:19.854942140Z         keras_module = keras_utils.keras()
2020-03-22T18:26:19.854944273Z         floatx = keras_module.backend.floatx()
2020-03-22T18:26:19.854946321Z     
2020-03-22T18:26:19.854948304Z         custom_objects = self.getCustomObjects()
2020-03-22T18:26:19.854950368Z     
2020-03-22T18:26:19.854952340Z >       history, serialized_model, hvd_size = run_results[0]
2020-03-22T18:26:19.854957386Z E       TypeError: 'NoneType' object is not iterable
```

and

```
2020-03-22T18:24:50.377254213Z >               raise RuntimeError("mpirun failed with exit code {exit_code}".format(exit_code=exit_code))
2020-03-22T18:24:50.377257104Z E               RuntimeError: mpirun failed with exit code 134
```

**Comment by EnricoMi** [comment]
@tgaddair I have deactivated Spark Estimators tests for GPU and Gloo (see [fcb814a](https://github.com/horovod/horovod/pull/1807/commits/fcb814a6fa447ad53bdfa0ef264af3c5b495dbfa)) and will handle this in a separate PR. I will need your input though, I suspect (see [comment above](https://github.com/horovod/horovod/pull/1807#issuecomment-602501879)).

**Comment by tgaddair** [comment]
This is awesome work @EnricoMi!  I think the test failures are consistent with some other issues we saw recently.  I agree with waiting for a followup PR to investigate.  I'll take some time to review this throughly today (may spill into tomorrow).  Thanks!

**Comment by EnricoMi** [comment]
Hope this helps to understands the mechanics of Spark over Gloo:
spark-gloo

File: test/test_run.py [hunk 1]
```diff
@@ -245,6 +222,132 @@ def test_host_hash(self):
...
+                mpi_run.assert_called_once()
+                js_run.assert_not_called()
+            elif expected == "js":
+                gloo_run.assert_not_called()
+                mpi_run.assert_not_called()
+                js_run.assert_called_once()
+            else:
+                raise ValueError("unsupported framework: {}".format(expected))
+
+        for use_gloo in [None, False, True]:
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply test/test_run.py]: fixed, great trick

File: test/test_run.py [hunk 2]
```diff
@@ -245,6 +222,132 @@ def test_host_hash(self):
...
+
+        test(("mpirun (Open MPI) 2.1.1\n"
+              "\n"
+              "Report bugs to http://www.open-mpi.org/community/help/\n"), _OMPI_IMPL)
+
+        test("OpenRTE", _OMPI_IMPL)
+
+        test("IBM Spectrum MPI", _SMPI_IMPL)
+
+        test(("HYDRA build details:\n"
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply test/test_run.py]: I though I'd rather test with a real string, but I admit I don't like it either. Truncated it a bit.

File: test/test_interactiverun.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 
 import horovod.torch as hvd
 
-from horovod.run import run
+from horovod.run.run import run
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply test/test_interactiverun.py]: This is a tricky one:
in `test_run` I want to patch the `gloo_built` and `mpi_built` (done in test/common.py#121-122) that are imported in `horovod/run/run.py`. But `horovod/run/__init__.py` imports `run()` from `run.py` as `horovod.run.run`, so I cannot reference `horovod.run.run.gloo_built` anymore.

**Comment by EnricoMi** [comment]
One caveat with Spark over Gloo is that you cannot see your training functions stdout / stderr anymore. MPI showed this nicely with node/slot prefix on our driver's stdout / stdin. With gloo those outputs are only visible in the workers' stdout / stderr. Those are available in the Spark UI though.

File: test/test_interactiverun.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgaddair) [reply test/test_interactiverun.py]: Easiest solution would be to rename `run.py` to `runner.py`.  That way we don't change the public API, only internals.

However, I do think we should ultimately change the public API on this one, but in the other direction, to:

```
import horovod
horovod.run(...)
```

That's a bigger refactor, though, so I'd rather do that in a separate PR.

**Comment by tgaddair** [comment]
> One caveat with Spark over Gloo is that you cannot see your training functions stdout / stderr anymore. MPI showed this nicely with node/slot prefix on our driver's stdout / stderr. With gloo those outputs are only visible in the workers' stdout / stderr. Those are available in the Spark UI though.

I see.  I'd say that's fine for now.  I have some thoughts on how we could implement this, but this is already a substantial PR.  Let's add stdout/stderr forwarding the a follow-up.

**Comment by EnricoMi** [comment]
> I see. I'd say that's fine for now. I have some thoughts on how we could implement this, but this is already a substantial PR. Let's add stdout/stderr forwarding the a follow-up.

My thoughts!

File: test/test_interactiverun.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EnricoMi) [reply test/test_interactiverun.py]: Agree, renaming `run.py` is minimal if `horovod.run.runner` is not public API.

File: horovod/run/common/util/tiny_shell_exec.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+import six
+import sys
+import traceback
+
+from horovod.run.common.util import safe_shell_exec
+
+
+def execute(command):
+    """
```
Comment (abditag2) [inline horovod/run/common/util/tiny_shell_exec.py]: Is there any reason for not using `safe_shell_exec.py`?

File: test/test_interactiverun.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EnricoMi) [reply test/test_interactiverun.py]: I have also renamed `horovod/spark/run.py` to align with `horovod.run.runner`.

**Comment by abditag2** [comment]
@EnricoMi Thanks for the great work!

File: horovod/run/common/util/tiny_shell_exec.py (hunk 4 above)
Comment (EnricoMi) [reply horovod/run/common/util/tiny_shell_exec.py]: I have extracted this code from `_get_mpi_implementation_flags` to make that method testable. Didn't want to refactor the how `_get_mpi_implementation_flags` used `safe_shell_exec.execute(...)` internally.

**Pull request merged**